### ISSUE TITLE
ユーザー指定による自由な optim & scheduler 自動登録と使用

### ISSUE THREAD

**Issue by muooon** [iron-mukakin/Animerge#3]
新しい学習ツールをありがとうございます。kohya-script ベースの信頼性と、GUIによる簡単操作は誰もが望むツールです。まとめあげるのは相当に大変だったろうと思います。現状 Addift などのタブがなく、どこで指定をするかはわからないのですが、表題の件について提案させてください。
それと、LoRA学習のユーザー指定について、プリセット対応だけでなく、前回値をキャッシュすると利便性を向上させるかなと思います。これはLoRA学習だけでなく、マージなどでも同様にキャッシュあると便利かなと思います。ご検討ください。

対象：lora_train.py
効果：optim & scheduler についてユーザーの自由追加を常時可能にする
(新規追加のコード変更作業を極力減らせると思います)
方法：｢sd-scripts｣の直下に、｢optimizer｣、｢scheduler｣、フォルダを作成し、それぞれに __init__ 、及び optim & scheduler のファイルを配置する

変更部分抜粋 (最初のものにエラーあったため修正しました)

```

from typing import Callable
import importlib
import inspect
import pkgutil
from torch.optim import Optimizer


def discover_custom_optimizers(sd_scripts_root):
    """sd-scripts/optimizer/ 以下のOptimizerクラスを自動検出する。"""
    optimizers = {}
    if str(sd_scripts_root) not in sys.path:
        sys.path.insert(0, str(sd_scripts_root))

    try:
        import optimizer as optimizer_package
    except ImportError:
        return optimizers

    for module_info in pkgutil.iter_modules(optimizer_package.__path__):
        module_name = module_info.name

        # __init__ 等は除外
        if module_name.startswith("_"):
            continue

        try:
            module = importlib.import_module(
                f"optimizer.{module_name}"
            )
        except Exception:
            continue

        for name, obj in inspect.getmembers(module, inspect.isclass):
            # 外部からimportされているOptimizerクラスを誤登録しない
            if obj.__module__ != module.__name__:
                continue

            if not issubclass(obj, Optimizer):
                continue

            optimizers[name] = f"optimizer.{module_name}.{name}"

    return optimizers

def discover_custom_schedulers(sd_scripts_root: Path):
    """sd-scripts/scheduler/ 以下のSchedulerを自動検出する。"""
    schedulers = {}

    scheduler_dir = sd_scripts_root / "scheduler"
    if not scheduler_dir.is_dir():
        return schedulers

    if str(sd_scripts_root) not in sys.path:
        sys.path.insert(0, str(sd_scripts_root))

    try:
        import scheduler as scheduler_package
    except ImportError:
        return schedulers

    for module_info in pkgutil.iter_modules(scheduler_package.__path__):
        module_name = module_info.name

        if module_name.startswith("_"):
            continue

        try:
            module = importlib.import_module(
                f"scheduler.{module_name}"
            )
        except Exception:
            continue

        for name, obj in inspect.getmembers(module, inspect.isclass):
            if obj.__module__ != module.__name__:
                continue

            schedulers[name] = f"scheduler.{module_name}.{name}"

    return schedulers


# ──────────────────────────────────────────────────────────────────────────────
# 定数 (anima_train_utils.py / anima_train_network.py 由来)
# ──────────────────────────────────────────────────────────────────────────────
SD_SCRIPTS_ROOT = Path(__file__).resolve().parent.parent / "sd-scripts"
OPTIMIZER_TYPES = discover_custom_optimizers(SD_SCRIPTS_ROOT)
SCHEDULER_TYPES = discover_custom_schedulers(SD_SCRIPTS_ROOT)

OPTIMIZERS = [
    "AdamW", "AdamW8bit", "Adafactor", "DAdaptAdam",
    "DAdaptAdaGrad", "DAdaptSGD", "Lion", "Prodigy",
]

for name in OPTIMIZER_TYPES:
    if name not in OPTIMIZERS:
        OPTIMIZERS.append(name)

LR_SCHEDULERS = [
    "constant", "constant_with_warmup", "cosine",
    "cosine_with_restarts", "linear", "polynomial",
]

for name in SCHEDULER_TYPES:
    if name not in LR_SCHEDULERS:
        LR_SCHEDULERS.append(name)

PRECISIONS = ["bf16", "fp16", "fp32"]
TIMESTEP_SAMPLING = ["sigmoid", "sigma", "uniform", "shift", "flux_shift"]
ATTN_MODES = ["torch", "xformers", "flash", "sdpa"]
WEIGHTING_SCHEMES = ["none", "sigma_sqrt", "cosmap"]

# 階層学習 GUI定数（gui.py の MATRIX_BLOCKS / MATRIX_COMPONENTS / COMPONENT_GROUPS と同一）
LAYER_TRAIN_MODES   = ("Matrix", "Transformer", "Component")
MATRIX_BLOCKS       = ("Input", "Middle", "Output")
MATRIX_COMPONENTS   = ("Attention", "MLP", "Norm", "ResNet", "Timestep")
COMPONENT_GROUPS    = ("Attention", "MLP", "Norm", "ResNet", "Timestep", "Other")
LAYER_COLUMNS       = 3
SAMPLE_FIXED_SEED   = 42

---

def _build_command(s: _TrainState) -> list[str]:
    """GUIの設定値から accelerate launch コマンドリストを生成する。"""
    sd_scripts_root = s.paths.root / "sd-scripts"
    train_script = sd_scripts_root / "anima_train_network.py"

    # accelerate launch はspawnするためPYTHONPATHが引き継がれない場合がある。
    # ラッパー経由でsys.pathを確実に追加する。
    wrapper = sd_scripts_root / "_gui_train_wrapper.py"
    wrapper.write_text(
        "import sys, os\n"
        f"sys.path.insert(0, r'{sd_scripts_root}')\n"
        f"os.chdir(r'{sd_scripts_root}')\n"
        f"with open(r'{train_script}', encoding='utf-8') as _f:\n"
        "    _code = compile(_f.read(), _f.name, 'exec')\n"
        f"exec(_code, {{'__name__': '__main__', '__file__': r'{train_script}'}})\n",
        encoding="utf-8",
    )

    cmd = [
        sys.executable, "-m", "accelerate.commands.launch",
        "--mixed_precision", s.mixed_precision.get(),
        "--num_cpu_threads_per_process", "1",
        str(wrapper),
        "--pretrained_model_name_or_path", s.model_path.get(),
        "--vae",           s.vae_path.get(),
        "--qwen3",         s.qwen3_path.get(),
        "--train_data_dir", s.train_data_dir.get(),
        "--output_dir",    s.output_dir.get(),
        "--output_name",   s.output_name.get(),
        "--resolution",    s.resolution.get(),
        "--train_batch_size", str(s.batch_size.get()),
        "--caption_extension", s.caption_extension.get(),
        "--keep_tokens",   str(s.keep_tokens.get()),
        "--network_module", s.network_module.get(),
        "--network_dim",   str(s.network_dim.get()),
        "--network_alpha", str(s.network_alpha.get()),
        "--learning_rate", s.lr.get(),
        "--lr_warmup_steps", str(s.lr_warmup_steps.get()),
        "--optimizer_type", OPTIMIZER_TYPES.get(
            s.optimizer.get(),
            s.optimizer.get()
        ),
        "--max_train_epochs", str(s.max_train_epochs.get()),
        "--save_every_n_epochs", str(s.save_every_n_epochs.get()),
        "--mixed_precision", s.mixed_precision.get(),
        "--save_precision", s.precision.get(),
        "--gradient_accumulation_steps", str(s.grad_accum.get()),
        "--max_grad_norm", str(s.max_grad_norm.get()),
        "--timestep_sampling", s.timestep_sampling.get(),
        "--discrete_flow_shift", str(s.discrete_flow_shift.get()),
        "--sigmoid_scale", str(s.sigmoid_scale.get()),
        "--weighting_scheme", s.weighting_scheme.get(),
        "--attn_mode",     s.attn_mode.get(),
        "--qwen3_max_token_length", str(s.qwen3_max_token_length.get()),
        "--t5_max_token_length", str(s.t5_max_token_length.get()),
        "--max_data_loader_n_workers", "0",
    ]

    scheduler_name = s.lr_scheduler.get()

    if scheduler_name in SCHEDULER_TYPES:
        cmd += [
            "--lr_scheduler_type",
            SCHEDULER_TYPES[scheduler_name],
        ]
    else:
        cmd += [
            "--lr_scheduler",
            scheduler_name,
        ]

    # オプション引数
    if s.seed.get():
        cmd += ["--seed", s.seed.get()]
    if s.llm_adapter_path.get():
        cmd += ["--llm_adapter_path", s.llm_adapter_path.get()]
    if s.network_weights.get():
        cmd += ["--network_weights", s.network_weights.get()]
    if s.optimizer_args.get():
        cmd += ["--optimizer_args"] + s.optimizer_args.get().split()
    if s.t5_tokenizer_path.get():
        cmd += ["--t5_tokenizer_path", s.t5_tokenizer_path.get()]


```

**Comment by iron-mukakin** [comment]
コードありがとうございます。GUIの入力値キャッシュと一緒にマージ調整してみます。
